fedimint-api/src/db/mod.rs

```diff
@@ -71,6 +96,64 @@ impl Database {
     pub async fn begin_transaction(&self) -> DatabaseTransaction {
         DatabaseTransaction::new(self.0.db.begin_transaction().await, &self.0.module_decoders)
     }
+
+    /// Runs a closure with a reference to a database transaction and tries to commit the
+    /// transaction if the closure returns `Ok` and rolls it back otherwise. If committing fails the
+    /// closure is run again for up to `retries` times. If retries is `None` it will run
+    /// `usize::MAX` times which is close enough to infinite times.
+    ///
+    /// The closure `tx_fn` provided should not have side effects outside of the database
+    /// transaction provided, or if it does these should be idempotent, since the closure might be
+    /// run multiple times.
+    ///
+    /// # Lifetime Parameters
+    ///
+    /// The higher rank trait bound (HRTB) `'a` that is applied to the the mutable reference to the
+    /// database transaction ensures that the reference lives as least as long as the returned
+    /// future of the closure.
+    ///
+    /// Further, the reference to self (`'s`) must outlive the `DatabaseTransaction<'dt>`. In other
+    /// words, the `DatabaseTransaction` must live as least as long as `self` and that is true as
+    /// the `DatabaseTransaction` is only dropped at the end of the `loop{}`.
+    pub async fn autocommit<'s: 'dt, 'dt, F, T, E>(
+        &'s self,
+        tx_fn: F,
+        max_retries: Option<usize>,
+    ) -> Result<T, AutocommitError<E>>
+    where
+        for<'a> F: Fn(
+            &'a mut DatabaseTransaction<'dt>,
+        ) -> Pin<Box<dyn Future<Output = Result<T, E>> + 'a>>,
+    {
+        let mut retries = 0;
+        loop {
+            let mut dbtx = self.begin_transaction().await;
+
+            match tx_fn(&mut dbtx).await {
+                Ok(val) => {
+                    match dbtx.commit_tx().await {
+                        Ok(()) => {
+                            return Ok(val);
+                        }
+                        Err(e) if max_retries.map(|mr| mr <= retries).unwrap_or(false) => {
+                            return Err(AutocommitError::CommitFailed {
+                                retries,
+                                last_error: e,
+                            });
+                        }
+                        Err(_) => {
+                            // try again
+                        }
+                    }
+                }
+                Err(e) => {
+                    return Err(AutocommitError::ClosureError { retries, error: e });
+                }
+            };
+
+            retries += 1;
```

Comment:
This might overflow after `usize::MAX` rounds. For `usize=u32` that could potentially happen (1ms per iteration would mean it happens after 50 days), for `usize=u64` probably not. Then again, if your federation is stuck for so long trying to commit a DB tx, do you really care?